### PR TITLE
Add `font-size` to theme

### LINKED ISSUE #1417: Allow scaling of font to accommodate hidpi systems
On my hidpi system, Napari's fonts are very small.

Ideally, one would be able to zoom the interface by typing, e.g., Ctrl-+, but that is not currently implemented.

As a start, I would like to ask whether we can replace absolute font-sizes (and perhaps other sizes) in the .qss files with relative sizes (`em` instead of `pt` or `px`).  With this in place, I'd only have to add a single specifier to `02_custom.qss` to increase app-wide font-size.

There is a note in `00_base.qss` already saying:

> might be possible to convert px to em by 1px = 0.0625em

but I'm not sure where that came from.

If you think this is a reasonable request, I can poke around and see if I can make it work.

### PR DESCRIPTION
**Pull request by lukasz-migas**
# Fixes/Closes

closes #1417 

# Description

This PR adds a new `font_size` attribute to the theme which is injected sporadically throughout the stylesheets.

A fix for an issue on Windows (perhaps other OSs too?) that have multiple high-res screens connected (and use PySide2 as the backend) was done initially here but was moved to its own PR. See PR #5968 

Before:
- text in the app is way too small to use
image

After:
- text is of expected size and can be changed by changing the `.font_size` attribute of the current theme
image

Some problems:
Before the change, when starting `napari`, I see no warnings in the terminal but after running from this branch, I get the following warnings and have no idea where/why its appearing:

`
WARNING: DirectWrite: CreateFontFaceFromHDC() failed (Indicates an error in an input file such as a font file.) for QFontDef(Family="", pointsize=12, pixelsize=16, styleHint=5, weight=50, stretch=100, hintingPreference=0) LOGFONT("MS Sans Serif", lfWidth=0, lfHeight=-16) dpi=192
14:09:30 WARNING DirectWrite: CreateFontFaceFromHDC() failed (Indicates an error in an input file such as a font file.) for QFontDef(Family="", pointsize=12, pixelsize=16, styleHint=5, weight=50, stretch=100, hintingPreference=0) LOGFONT("MS Sans Serif", lfWidth=0, lfHeight=-16) dpi=192
`

Edit:  Maybe the warning was related with the font of the `napari-console` not being properly set

Using the evented model:
https://user-images.githubusercontent.com/25161821/223434082-eaeb2dbd-dba5-472e-a2bf-54a2a6c698ba.mp4

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- [x] all tests pass with my change
- [x] I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 1]
```diff
@@ -182,7 +182,7 @@ QtLayerWidget > QLineEdit {
   border: none;
   border-radius: 2px;
   padding: 2px;
-  font-size: 14px;
+  font-size: {{ increase(font_size, 2) }};
```
Comment (Carreau) [inline napari/_qt/qt_resources/styles/02_custom.qss:186]: Sorry, stupid question maybe but this is/was 14px, while the other places were in `pt`, does that change anything ?

It's here and in a bunch of places, so I don't know if it's relevant or not.

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 2]
```diff
@@ -371,7 +371,7 @@ QtDimSliderWidget > QLineEdit {
 
 #QtPopupFrame > QLabel {
   color: {{ darken(text, 35) }};
-  font-size: 12px;
+  font-size: {{ font_size }};
 }
```
Comment (Carreau) [inline napari/_qt/qt_resources/styles/02_custom.qss:376]: there is pt/px difference here as well

File: napari/_qt/qt_main_window.py [hunk 3]
```diff
@@ -1271,12 +1273,8 @@ def _update_theme(self, event=None):
...
             if value == "system":
                 # system isn't a theme, so get the name and set style sheet
-                actual_theme_name = get_system_theme()
-                self._qt_window.setStyleSheet(
-                    get_stylesheet(actual_theme_name)
-                )
-            else:
-                self._qt_window.setStyleSheet(get_stylesheet(value))
+                value = get_system_theme()
+            self._qt_window.setStyleSheet(get_stylesheet(value))
```
Comment (Carreau) [inline napari/_qt/qt_main_window.py]: +1

File: napari/_qt/qt_event_loop.py [hunk 4]
```diff
@@ -139,13 +138,10 @@ def get_app(
...
-            QApplication.setAttribute(
-                Qt.ApplicationAttribute.AA_EnableHighDpiScaling
-            )
-            QApplication.setAttribute(
-                Qt.ApplicationAttribute.AA_UseHighDpiPixmaps
-            )
+        QApplication.setAttribute(
+            Qt.ApplicationAttribute.AA_EnableHighDpiScaling
+        )
+        QApplication.setAttribute(Qt.ApplicationAttribute.AA_UseHighDpiPixmaps)
```
Comment (Carreau) [inline napari/_qt/qt_event_loop.py]: Do you have any idea why there was a if PYQT5 and it's not necessary anymore ? Just to make sure.

File: napari/utils/theme.py [hunk 5]
```diff
@@ -91,13 +94,35 @@ def _ensure_syntax_style(value: str) -> str:
         )
         return value
 
+    @validator("font_size", pre=True)
+    def _ensure_font_size(value: str) -> str:
+        assert "pt" in value, trans._(
```
Comment (Carreau) [inline napari/utils/theme.py]: Suggestion to be stricter on where pt is ? I assume if it's not at the end it will fail anyway so up to you.
```suggestion
        assert value.endswith('pt'), trans._(
```

**Review by Carreau — APPROVED** [review]
I have some vague questions for my own understanding, but nothing blocking +1 to make things more consistent

File: napari/_qt/qt_resources/styles/02_custom.qss (hunk 1 above)
Comment (Czaki) [reply napari/_qt/qt_resources/styles/02_custom.qss:186]: We should everywhere use pt. `px` is based on physical screen pixel size, and `pt` base on dpi of the screen.

File: napari/_qt/qt_event_loop.py (hunk 4 above)
Comment (Carreau) [reply napari/_qt/qt_event_loop.py]: The conditional seem to be actually needed as on pyqt6 the tests fails with:


```
  Traceback (most recent call last):
    File "<string>", line 4, in <module>
    File "/home/runner/work/napari/napari/napari/view_layers.py", line 179, in view_image
      return _make_viewer_then('add_image', *args, **kwargs)[0]
    File "/home/runner/work/napari/napari/napari/view_layers.py", line 157, in _make_viewer_then
      viewer = Viewer(**vkwargs)
    File "/home/runner/work/napari/napari/napari/viewer.py", line 67, in __init__
      self._window = Window(self, show=show)
    File "/home/runner/work/napari/napari/napari/_qt/qt_main_window.py", line 555, in __init__
      get_app()
    File "/home/runner/work/napari/napari/napari/_qt/qt_event_loop.py", line 141, in get_app
      Qt.ApplicationAttribute.AA_EnableHighDpiScaling
    File "/opt/hostedtoolcache/Python/3.9.16/x64/lib/python3.9/enum.py", line 429, in __getattr__
      raise AttributeError(name) from None
  AttributeError: AA_EnableHighDpiScaling
```

File: napari/_qt/qt_resources/styles/02_custom.qss (hunk 1 above)
Comment (Carreau) [reply napari/_qt/qt_resources/styles/02_custom.qss:186]: > We should everywhere use pt. px is based on physical screen pixel size, and pt base on dpi of the screen.

Sure I know the pt/px/em/ex units, let me rephrase.

Is replacing 14px by 14pt the same for qt or do we need to have a correcting factor. Or maybe it was just a typo and was 14pt what was originally meant.

It also relatively easy to scan through a file and not realize that some units are pt and px.

File: napari/_qt/qt_event_loop.py (hunk 4 above)
Comment (Czaki) [reply napari/_qt/qt_event_loop.py]: Yes. It was removed in Qt6. 
But I do not know/remeber why it was introduced.

File: napari/_qt/qt_main_window.py (hunk 3 above)
Comment (psobolewskiPhD) [reply napari/_qt/qt_main_window.py]: So this changes the theme behavior, such that viewer.theme only returns `light` or `dark`.
I'm actually supportive of this, but when I implemented the disconnect between the settings and the current viewer (#5143), it was contentious, with the consensus being that viewer.theme should report `system` if that was indeed the theme set and that the stylesheet should be adjusted appropriately. By letting viewer.theme retain `system`, it may help future serialization of the viewer state or let the theme adapt to the OS theme.
See these two comments:
https://github.com/napari/napari/pull/5143#issuecomment-1336213640
https://github.com/napari/napari/pull/5143#issuecomment-1342178406

I'm not sure it's needed to change this behavior to implement the changes.

File: napari/_qt/qt_event_loop.py (hunk 4 above)
Comment (psobolewskiPhD) [reply napari/_qt/qt_event_loop.py]: Qt::AA_EnableHighDpiScaling was added in Qt5 (5.6) and has been removed in Qt6.2 (https://doc.qt.io/qtforpython/porting_from2.html)
So I think this does need to be within an `If` so we can (experimentally) support Qt6.

**Review by psobolewskiPhD — CHANGES_REQUESTED** [review]
For me (macOS) the font size with this PR is smaller in plugin widgets (Light is live, dark is this PR)
image
image

Also, if I edit `theme.py` to change the number to something like 32 or 40, nothing happens?

Further, this isn't just change to font size, but also the font itself in the napari-console.
With this PR:
image
Note it's smaller and a sans-serif that is not monospace.

For reference, without:
image

(This can actually be seen in the screenshots in OP)

**Comment by kephale** [comment]
@lukasz-migas Just letting you know, we're looking into wrapping up this PR.

@dalthviz would you be able to check this out?

File: napari/_qt/qt_main_window.py (hunk 3 above)
Comment (andy-sweet) [reply napari/_qt/qt_main_window.py]: Agreed with @psobolewskiPhD that the system and light/dark discussion seems separate enough, that we should think about it in a different PR.

**Comment by lukasz-migas** [comment]
@kephale Hi - sorry that I've not finished this PR yet. I am unable at the moment (conferences + work) but I can take a look at it after 10th June.

Sorry for the delay..

**Comment by kephale** [comment]
@lukasz-migas no worries at all! I just wanted to give you a heads up that folks from the community looked at your PR recently (at the PR party last week).

**Comment by dalthviz** [comment]
Hi @kephale and @lukasz-migas  ! I think I can help here if that is okay, let me know!

**Comment by kephale** [comment]
@lukasz-migas, would you mind if @dalthviz wraps this up? No worries if you'd like to do it yourself. Otherwise, we would just build on your PR anyway!

File: napari/_qt/qt_event_loop.py (hunk 4 above)
Comment (dalthviz) [reply napari/_qt/qt_event_loop.py]: Checking this, maybe what is needed is to change the validation to use `QT5` or something like `PYQT5 or PYSIDE2` instead of only the `PYQT5` constant? In that way the attributes will be applied for PyQt5 and PySide2 where they exists and are needed since the attributes are deactivated by default

File: napari/_qt/qt_resources/styles/02_custom.qss (hunk 1 above)
Comment (dalthviz) [reply napari/_qt/qt_resources/styles/02_custom.qss:186]: Checking this, I think there is a factor we will need to take into account since `pt` vs `px` is not a one on one equivalence. 14pt are actually ~18px (online converter to check: https://pixelsconverter.com/pt-to-px). Also, when checking this locally on Windows I can see that some of the aplication text has increased in size which probably is due to the changing to use pt but keeping the same px value:

### This PR:

imagen

imagen

### main

imagen

imagen

File: napari/_qt/qt_event_loop.py (hunk 4 above)
Comment (psobolewskiPhD) [reply napari/_qt/qt_event_loop.py]: Makes sense to me.

File: napari/_qt/qt_resources/styles/02_custom.qss (hunk 1 above)
Comment (psobolewskiPhD) [reply napari/_qt/qt_resources/styles/02_custom.qss:186]: I think the px/pts relation also depends on the `ppi` resolution of the display, which differs between windows and macOS, so that's probably the effect.

File: napari/_qt/qt_main_window.py (hunk 3 above)
Comment (dalthviz) [reply napari/_qt/qt_main_window.py]: Checking this change, I think the value of `viewer.theme` still can be `system`. Seems to me that the change here is just a little refactor to not have to call twice the `self._qt_window.setStyleSheet` while reusing the `value` variable, no? Testing locally, if I selected `system` as theme value I can still see from the IPython Console that `viewer.theme` outputs `system`:

imagen

Comment (dalthviz) [reply napari/_qt/qt_main_window.py]: Maybe preserving the use of the variable `actual_theme_name` could make things more clear here? So something like:

```python
        actual_theme_name = value    
        if value == "system":
            # system isn't a theme, so get the name and set style sheet
            actual_theme_name = get_system_theme()
        self._qt_window.setStyleSheet(get_stylesheet(actual_theme_name))

```

Comment (psobolewskiPhD) [reply napari/_qt/qt_main_window.py]: Gah, you're right! I mis-read the code. It's actually refactoring @Czaki suggested in the original PR and I didn't do because I thought it wasn't clear what was going on.

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 6]
```diff
@@ -524,23 +524,23 @@ NapariQtNotification #close_button {
 
 NapariQtNotification #source_label {
   color: {{ primary }};
-  font-size: 11px;
+  font-size: {{ decrease(font_size, 1) }};
```
Comment (dalthviz) [inline napari/_qt/qt_resources/styles/02_custom.qss:528]: px value to check

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 7]
```diff
@@ -524,23 +524,23 @@ NapariQtNotification #close_button {
...
 }
 
 NapariQtNotification #severity_icon {
   padding: 0;
   margin: 0 0 -3px 0;
   min-width: 20px;
   min-height: 18px;
-  font-size: 15px;
+  font-size: {{ increase(font_size, 3) }};
```
Comment (dalthviz) [inline napari/_qt/qt_resources/styles/02_custom.qss:536]: px value to check

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 8]
```diff
@@ -623,12 +623,12 @@ QPushButton#close_button:disabled {
 
 #small_text {
   color: {{ opacity(text, 150) }};
-  font-size: 10px;
+  font-size: {{ decrease(font_size, 2) }};
```
Comment (dalthviz) [inline napari/_qt/qt_resources/styles/02_custom.qss:671]: px value to check

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 9]
```diff
@@ -623,12 +623,12 @@ QPushButton#close_button:disabled {
...
 }
 
 #small_italic_text {
   color: {{ opacity(text, 150) }};
-  font-size: 12px;
+  font-size: {{ font_size }};
```
Comment (dalthviz) [inline napari/_qt/qt_resources/styles/02_custom.qss:676]: px value to check

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 10]
```diff
@@ -728,12 +728,12 @@ QtWelcomeWidget[drag=true] {
 
 QtWelcomeLabel {
  color: {{ foreground }};
- font-size: 20px;
+ font-size: {{ increase(font_size, 8) }};
```
Comment (dalthviz) [inline napari/_qt/qt_resources/styles/02_custom.qss:776]: px value to check

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 11]
```diff
@@ -728,12 +728,12 @@ QtWelcomeWidget[drag=true] {
...
 }
 
 QtShortcutLabel {
  color: {{ foreground }};
- font-size: 16px;
+ font-size: {{ increase(font_size, 4) }};
```
Comment (dalthviz) [inline napari/_qt/qt_resources/styles/02_custom.qss:781]: px value to check

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 12]
```diff
@@ -789,7 +789,7 @@ QtListView::item:selected:!active {
 QtListView QLineEdit {
   background-color: {{ darken(current, 20) }};
   selection-background-color: {{ lighten(current, 20) }};
-  font-size: 12px;
+  font-size: {{ font_size }};
```
Comment (dalthviz) [inline napari/_qt/qt_resources/styles/02_custom.qss:837]: px value to check

File: napari/_qt/qt_resources/styles/02_custom.qss (hunk 1 above)
Comment (dalthviz) [reply napari/_qt/qt_resources/styles/02_custom.qss:186]: Oh I see, so should we create some function that helps with the conversion from px to pt and that queries the current value of `ppi`? 

Or maybe define some px to pt baseline equivalence to use in order to replace the px values that are being changed here and that work for Windows, MacOs and Linux? Looking around the PR seems like the values in px were:

* 10px
* 11px
* 12px
* 14px 
* 15px
* 16px
* 20px

Comment (dalthviz) [reply napari/_qt/qt_resources/styles/02_custom.qss:186]: Also, should the default `font-size` be set to 12pt as it is right now (and I guess though as equivalent of 12px) or a different value? 🤔

File: napari/utils/theme.py (hunk 5 above)
Comment (dalthviz) [reply napari/utils/theme.py]: Makes sense to me :+1:

File: napari/_qt/qt_resources/styles/00_base.qss [hunk 13]
```diff
@@ -57,6 +57,7 @@ QTextEdit {
   color: {{ text }};
   selection-background-color: {{ foreground }};
   padding: 4px 2px 4px 4px;
+  font-size: {{ font_size }};
```
Comment (dalthviz) [inline napari/_qt/qt_resources/styles/00_base.qss]: Seems like this line is what is causing the `napari-console` to change its font 🤔

File: napari/_qt/qt_resources/styles/02_custom.qss (hunk 1 above)
Comment (dalthviz) [reply napari/_qt/qt_resources/styles/02_custom.qss:186]: So, after playing a little bit with the base theme value, seems like using 13pt on macOS and 9pt on Windows and Linux preserves how things look on main. However extra checking from others is appreciated!

Edit: Leaving 12pt on macOS and tweaking some styles rules did the trick instead of using 13pt

**Comment by dalthviz** [comment]
I think the last commit should help with the widgets font size. Checked locally (Windows, Linux and macOS) and seems like now things are looking as if you were running main. Let me know if with the latest change things look better for you too @psobolewskiPhD ! Also, if someone else could check on Linux and Windows too that would be great :)

File: napari/_qt/qt_resources/styles/02_custom.qss [hunk 14]
```diff
@@ -302,12 +303,12 @@ QtColorBox {
 /* ----------------- QtLayerControls -------------------- */
 
 QtLayerControls > QLabel, QtLayerControls, QtPlaneControls > QLabeledSlider > QAbstractSpinBox {
-  font-size: 11pt;
+  font-size: {{ font_size }};
```
Comment (dalthviz) [inline napari/_qt/qt_resources/styles/02_custom.qss]: Seems like this should be `font-size: {{ decrease(font_size, 1) }};` (if using 12pt as base font size value)

**Comment by dalthviz** [comment]
Oh yes you are right! Will try changing the base font size value to be 12pt again on macOS and try to see if there are style rules that need to be modified. I think I found a rule that was no preserving the difference with 12pt so maybe there are other rules that need also a tweak :+1:

**Comment by dalthviz** [comment]
How things are looking for me locally with the latest commit:

| This PR | main |
|----|----|
| image | image |
| image | image |

I don't see differences anymore but maybe I'm missing some other little tweaks so any further checks are appreciated!

**Comment by dalthviz** [comment]
Note: Maybe this is related to https://github.com/napari/napari/issues/1417 ?

**Review by psobolewskiPhD — APPROVED** [review]
Everything makes sense and looks great on my end (macOS arm64, pyqt5 and pyside6).
Thanks @dalthviz !
And yes, I think this should close https://github.com/napari/napari/issues/1417
But I'm not sure what system was being used there.

I think this also opens up a path to providing `font-size` as a setting in Preferences, perhaps worth making a `Task` issue?

**Comment by psobolewskiPhD** [comment]
Maybe worth updating the theme example?
https://github.com/napari/napari/blob/main/examples/new_theme.py
I could see folks wanting larger font-size for docs screenshots, screenshare, etc. 

I *think* we need to make a task issue/PR for npe2 as well to make this available to plugins:
https://github.com/napari/npe2/blob/18c051462e713bd0e464ff655399818e26cce73e/src/npe2/manifest/contributions/_themes.py#L39

**Comment by dalthviz** [comment]
> I think this also opens up a path to providing font-size as a setting in Preferences, perhaps worth making a Task issue?

Makes sense to me :+1:

> Maybe worth updating the theme example?

Will do!

> I think we need to make a task issue/PR for npe2 as well to make this available to plugins:

So the task will be to add `font_size` as new `Field`, right? Not sure how the theme works for the plugins, but if the plugins also can set themes makes sense to make things coherent there :+1:

> Also, is this really a bugfix or an enhancement? I lean to the later (enhancement) and I'm leaning to 0.5.0 for the milestone.

That's a good question! Thinking about it, this PR is kind of both in the sense that adds improvements in the font-size handling via the theme (the enhancement part) but also fixes an issue with the High DPI attributes not being applied when using PySide2 since the validation was only covering PyQt5 (the bugfix part).

I guess that since most of the changes here are related with the enhancement part makes sense to label it as `enhancement` and makes sense to put it then in a minor release (like 0.5.0) :+1: 

Also, could it be worthy to move to an independent PR the bugfix part of this PR (the one liner change that adds `PYSIDE2` in the validation to apply the High DPI attributes)? In that way that fix can be part of the patch/bugfix release (0.4.18)

**Comment by psobolewskiPhD** [comment]
> Also, could it be worthy to move to an independent PR the bugfix part of this PR (the one liner change that adds `PYSIDE2` in the validation to apply the High DPI attributes)? In that way that fix can be part of the patch/bugfix release (0.4.18)

@Czaki what do you think? Seems reasonable to me to get the bugfix in, but lease the rest to 0.5?

**Comment by Czaki** [comment]
Please make separate PR with bugfix.

**Comment by psobolewskiPhD** [comment]
@dalthviz I think everything looks in order now for this?
Then, once merged:
1. we need an issue to expose the font-size theme value in the Preferences?
2. we change the npe2 PR from draft?

**Comment by dalthviz** [comment]
> @dalthviz I think everything looks in order now for this?
> Then, once merged:
> 
> *  we need an issue to expose the font-size theme value in the Preferences?
> *  we change the npe2 PR from draft?

I think is ready @psobolewskiPhD ! And the next steps sound good to me :+1:

**Comment by psobolewskiPhD** [comment]
Sorry this slipped through the cracks. Marking ready to merge!

**Comment by GenevieveBuckley** [comment]
> Then, once merged:
> 1. we need an issue to expose the font-size theme value in the Preferences?

New issue created here: https://github.com/napari/napari/issues/6109

> 2. we change the npe2 PR from draft?

Have pinged the draft PR here: https://github.com/napari/npe2/pull/300

3. Could someone leave a quick comment on https://github.com/napari/napari/issues/1417, explaining how users can manually adjust the `font_size` of their napari theme?

**Comment by dalthviz** [comment]
> 3. Could someone leave a quick comment on https://github.com/napari/napari/issues/1417, explaining how users can manually adjust the font_size of their napari theme?

Thank you for pushing this forward @GenevieveBuckley ! Left a comment explaining how to change the theme `font_size` config at https://github.com/napari/napari/issues/1417#issuecomment-1660608035
